src/shared_state/conflicts.md

```diff
@@ -0,0 +1,145 @@
+
+Lamport timestamps, for example, set the counter of the current message to one more than the previously greatest timestamp.
+If Alice and Bob both tried to repeatedly set the description for a group, it might play out like so:
+
+1. Alejo sets the description using a message (A) with an id `0`
+2. Binita receives this message, but wants to correct a typo, so she sends an update (B) with id `1`. Alejo can determine that this was sent and received in the right order, so this should supercede his own description rather than be treated as a conflict
+3. They both decide to add some emojis to the description, and do so in parallel (messages C and D), but with different emojis. Each believes their message to correctly have the id `2`
+4. Before either of the previous messages have been delivered, Binita sends one more update (E) to add another emoji using id `3`
+5. Alejo and Binita receive each others' messages:
+    * because C and D both have id `2`, Binita can conclude that Alejo was not aware of her message when he authored his
+    * Alejo can conclude the same of Binita's D, but cannot determine whether she was aware of his C when she sent E. Because of this uncertainty, they should both treat these messages as having been authored concurrently, resulting in a conflict
+6. Alejo decides to resolve this conflict, authoring a new description which includes the corrected typo and all emojis added by either of them. When Binita receives it will have clearly been sent after both C and E
+
+![graph demonstrating causal ordering](../images/shared_state/causal-ordering.svg)
+
+This method is popular both for its relative simplicity and the small amount of overhead that is introduced by including a counter in each message.
+It is somewhat ambiguous about exactly which messages have been received, however, it remains very effective in systems that require a high number of messages with minimal overhead.
+For these reasons, it is used by [Yjs](https://yjs.dev/), which will be discussed later in this chapter.
+
+Another popular approach which is more explicit but less compact involves the use of [cryptographic hash functions](https://en.wikipedia.org/wiki/Cryptographic_hash_function).
+Instead of a single counter, a list of hashes of the preceding messages which were known at the time the message was authored.
+This makes it possible to detect when a message has not been delivered or if message history has somehow been altered (although the latter probably isn't a concern with webxdc apps).
+Variations of this technique are applied in [Git](https://git-scm.com/), [BitTorrent](https://en.wikipedia.org/wiki/BitTorrent),
+[Matrix](https://en.wikipedia.org/wiki/Matrix_(protocol)), [Secure Scuttlebutt](https://en.wikipedia.org/wiki/Secure_Scuttlebutt),
+[IPFS](https://en.wikipedia.org/wiki/InterPlanetary_File_System), and many other protocols intended to solve a variety of problems.
+
```

Comment:
Not mentioning blockchains like bitcoin on purpose? It's a good example of chaining blocks together with a hash pointing to the previous block. But no biggie if left out.

Reply:
I had mentioned it during some intermediary version, but I may have accidentally dropped it when shuffling some content around